### ISSUE TITLE
crash for debugging mode

### ISSUE THREAD

**Issue by weiyuan-jiang** [GEOS-ESM/GEOSldas#414]
https://github.com/GEOS-ESM/GEOSldas/blob/449308bddec7d7dbe6fe813250fbc166eabd4e04/src/Components/GEOSldas_GridComp/Shared/my_matrix_functions.F90#L749-L750

During debugging, Line 749 crashes with the message: 
(33): Shape mismatch: The extent of dimension 1 of array A is 4 and the corresponding extent of array A_STEP1 is 1

So I added two prints before Line 749:

print*, "nint(A_step2(1:N_step2,1)):", nint(A_step2(1:N_step2,1))
print*, "n_rows, n_step2:", n_rows, N_step2

Here is the print out

 nint(A_step2(1:N_step2,1)):           1
 n_rows, n_step2:           4           1

So there is mismatch.

**Comment by gmao-rreichle** [comment]
After staring at this for  a while, I'm wondering if lines 749-750 shouldn't rather read:

```
A(1:N_step2,1:2)      = A_step1(nint(A_step2(1:N_step2,1)),:)    
A(1:N_step2,  3)      = A_step2(             1:N_step2    ,2)
```
that is, with the added `1:N_step2` in the first dimension of A on the left-hand side.

Not sure how this was working before.  I suppose the compiler just threw random indices into the extra rows (beyond N_step2)? 

However, I don't understand why you get N_step2=1 in the first place.  I assume you only get to this place if you assimilate SMAP Tb.  For SMAP Tb, we usually have 4 so-called species: H-pol ascending, H-pol descending, V-pol ascending, V-pol descending.  The code in question looks for the unique species when ascending and descending are ignored. We should be left with 2 unique species: H-pol and V-pol.  Did you edit the `*ensupd*.nml` file?

In any case, there are lots of commented-out unit-test programs for the subroutine in question in the F90 file.  If we change anything, we need to make sure to thoroughly unit-test the subroutine.

**Comment by weiyuan-jiang** [comment]
This run is from night test.

**Comment by weiyuan-jiang** [comment]
Yes. We usually don't run debug mode in data assimilation

**Comment by weiyuan-jiang** [comment]
@gmao-rreichle 
The algorithm that calculates the unique rows seems ok. If there is a problem, it is the matrix passed in. 
  if (N_step2 ==1) then
      do i = 1,N_rows
        print*, "rows:", i, "data:", A(i,:)
      enddo
  endif

Results:
 rows:           1 data:   1.41000000E+09   40.0000000       2.00000000    
 rows:           2 data:   1.41000000E+09   40.0000000       2.00000000    
 rows:           3 data:   1.41000000E+09   40.0000000       2.00000000    
 rows:           4 data:   1.41000000E+09   40.0000000       2.00000000

**Comment by gmao-rreichle** [comment]
@weiyuan-jiang:
My mistake.  I didn't remember the details well enough and didn't look at the code surrounding the call to `unique_rows_3col( )` closely enough.  After reading the comments just ahead of the call I now realize again that H- and V-pol are also disregarded when identifying the unique rows.  This is because the RTM provides both H- and V-pol together anyway.  The three distinct criteria are (i) frequency, (ii) angle, and (iii) RTM ID (identifying the particular RTM physics to be used).  For SMAP, all species have the same values (ie, 1.41 GHz, 40 deg, and 2), so N_step2=1 is indeed what we want. 
In summary, we just need to explicitly use "1:N_step2" on the left-hand side as implemented in the branch.  
Let's keep this "issue" open until we have merged the associated PR #415
